src/Microsoft.AspNetCore.Routing/Tree/TreeRouter.cs

```diff
@@ -138,14 +138,18 @@ public VirtualPathData GetVirtualPath(VirtualPathContext context)
             // order. We just need to iterate them and use the first one that can generate a link.
             var matches = _linkGenerationTree.GetMatches(context);
 
-            foreach (var match in matches)
+            //Perf: Avoid allocation of enumerator if matches is empty
+            if(matches != null && matches.Count > 0 )
             {
-                var path = GenerateVirtualPath(context, match.Entry);
-                if (path != null)
+                foreach (var match in matches)
```

Comment:
We could use an indexer here since it's an `IList<T>`. Would remove an additional enumerator allocation.